rtcp.c

```diff
@@ -1044,3 +1044,345 @@ int janus_rtcp_nacks(char *packet, int len, GSList *nacks) {
 	rtcp->length = htons(words);
 	return words*4+4;
 }
+
+
+inline void set1(guint8 *data,size_t i,guint8 val)
+{
+	data[i] = val;
+}
+inline void set2(guint8 *data,size_t i,guint32 val)
+{
+	data[i+1] = (guint8)(val);
+	data[i]   = (guint8)(val>>8);
+}
+inline void set3(guint8 *data,size_t i,guint32 val)
+{
+	data[i+2] = (guint8)(val);
+	data[i+1] = (guint8)(val>>8);
+	data[i]   = (guint8)(val>>16);
+}
+inline void set4(guint8 *data,size_t i,guint32 val)
+{
+	data[i+3] = (guint8)(val);
+	data[i+2] = (guint8)(val>>8);
+	data[i+1] = (guint8)(val>>16);
+	data[i]   = (guint8)(val>>24);
+}
+
+enum PacketStatus
+{
+	NotReceived = 0,
+	SmallDelta = 1,
+	LargeOrNegativeDelta = 2,
+	Reserved = 3
+};
+		
+int janus_rtcp_transport_wide_cc_feedback(char *packet, size_t size, guint32 ssrc, guint32 media, guint8 feedback_packet_count, GQueue *transport_wide_cc_stats) {
+	if(packet == NULL || size < sizeof(janus_rtcp_header) || transport_wide_cc_stats == NULL || g_queue_is_empty(transport_wide_cc_stats))
+		return -1;
+	
+	memset(packet, 0, size);
+	janus_rtcp_header *rtcp = (janus_rtcp_header *)packet;
+	/* Set header */
+	rtcp->version = 2;
+	rtcp->type = RTCP_RTPFB;
+	rtcp->rc = 15;
+	/* Now set FB stuff */
+	janus_rtcp_fb *rtcpfb = (janus_rtcp_fb *)rtcp;
+	rtcpfb->ssrc = htonl(ssrc);
+	rtcpfb->media = htonl(media);
+	
+	/* Get first packet */
+	janus_rtcp_transport_wide_cc_stats* stat = (janus_rtcp_transport_wide_cc_stats*) g_queue_pop_head (transport_wide_cc_stats);
+	
+	/* Calculate temporal info */
+	guint16 base_seq_num	= stat->transport_seq_num;
+	gboolean first_received	= FALSE;
+	guint64 reference_time	= 0;
+	guint packet_status_count = g_queue_get_length(transport_wide_cc_stats) + 1;
+	
+	//JANUS_LOG(LOG_DBG, "base_seq_num:%u packet_status_count:%u feedback_packet_count:%u\n",base_seq_num, packet_status_count, feedback_packet_count);
+	
+	/*
+		0                   1                   2                   3
+		0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+	       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+	       |      base sequence number     |      packet status count      |
+	       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+	       |                 reference time                | fb pkt. count |
+	       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+	
+	 */
+	
+	/* The packet as unsigned */
+	guint8 *data = (guint8 *)packet;
+	/* The start of the feedback data */
+	size_t len = sizeof(janus_rtcp_header) + 8;
+
+	/* Set header data */
+	set2(data,len,base_seq_num);
+	set2(data,len+2,packet_status_count);
+	/* Set3 referenceTime when first received */
+	size_t reference_time_pos = len + 4;
+	set1(data,len+7,feedback_packet_count);
+	
+	/* Next byte */
+	len += 8;
+	
+	/* Initial time in us */
+	guint64 timestamp = 0;
+	
+	/* Store delta array */
+	GQueue* deltas = g_queue_new();
+	GQueue* statuses = g_queue_new();
```

Comment:
Still not sure about this additional allocations done for each RTCP feedback message, but I'll have to dig in the code deeper to understand if we can avoid those.